feat(plugins): re-org plugin docs; add frontend plugin guide (#2142)
spinnaker/spinnaker.github.io: guides/developer/plugins/frontend.md

---
layout: single
title: "Frontend Plugin Development"
sidebar:
  nav: guides
---

{% include fe-plugin-intro.md %}

{% include toc %}

# Required Tools

Before getting started make sure you have the following tools installed:

- [Gradle]
- [NPM]
- [Yarn]

[Gradle]: https://gradle.org/install/
[NPM]: https://docs.npmjs.com/cli/v7/configuring-npm/install
[Yarn]: https://yarnpkg.com/getting-started/install

> You will also need access to a running Spinnaker cluster `>= 1.20.6`.

Reviewer: ```suggestion
You need access to a running Spinnaker instance `>= 1.20.6`.
```

# Plugin Scaffolding

Frontend plugins can be written as part of a broader set of functionality or
on their own. In both cases you will need the Gradle toolchain to build and
release your plugin.

Here is what a typical plugin project will look like for a plugin named
`my-plugin`:

```shell
.
├── build.gradle
├── my-plugin-deck
├── my-plugin-gate
└── my-plugin-orca
```

The top-level `build.gradle` file contains metadata related to your plugin,
such as the plugin id that will allow you to add it to your Spinnaker cluster.

> Each core Spinnaker service will have it's own directory. This guide will focus

Reviewer 2: s/it's/its

Author: - [x] typo

your plugin and test in Deck, add all your stages to the `stage` field in your
plugin `index.ts` file like so:

```javascript
import { IDeckPlugin } from '@spinnaker/core';
import { widgetizeStage } from './WidgetizeStage';

export const plugin: IDeckPlugin = {
  stages: [widgetizeStage],
};
```

<!-- FIXME link to a frontend plugin example
https://github.com/spinnaker-plugin-examples/pf4jStagePlugin  needs to be updated to use the latest architecture
-->

## Overriding existing components

You can also override existing components within Deck, so long as they have
an `Overridable` annotation or are registered as an overridable component. In
this use case, you define your replacement component and then leverage the
`initialize` method of your plugin object to override the component when it is
loaded.

For example, if you want to remove the ability to modify Application
configuration in Deck, you define a component like this:

```javascript
import React from 'react';

export const InvisibleConfig = () => {
  return <h1>No config here!</h1>;
}
```

In order to override the component, you need to know the Application
configuration registration key. You can find its definition in the

> source project. There, you find the `Overridable` annotation is

Reviewer: which source project?

this use case, you define your replacement component and then leverage the
`initialize` method of your plugin object to override the component when it is
loaded.

For example, if you want to remove the ability to modify Application
configuration in Deck, you define a component like this:

```javascript
import React from 'react';

export const InvisibleConfig = () => {
  return <h1>No config here!</h1>;
}
```

In order to override the component, you need to know the Application
configuration registration key. You can find its definition in the
GitHub Deck project, [ApplicationConfig.tsx](https://github.com/spinnaker/deck/blob/43a0f56fa85fa1714fef0ba73810c90761f5996d/app/scripts/modules/core/src/application/config/ApplicationConfig.tsx). There, you find the `Overridable` annotation is
`applicationConfigView`. Then, in your `index.ts` file where you define your plugin object, you override that component in the `initialize` method. For example:

```javascript
import { IDeckPlugin, overrideRegistrationQueue } from '@spinnaker/core';
import { InvisibleConfig } from './InvisibleConfig';

export const plugin: IDeckPlugin = {

>   initialize: _ => {

Reviewer 2: I think `() => { ... }` is more idiomatic.

Author: - [x] stop treating TS like Haskell

accessible from a given Spinnaker cluster.

### Installation

Installation occurs either through [Halyard] or the [Spinnaker Operator]. In
either case you will have a Halconfig file that will configure plugins for your
Spinnaker cluster.

The following YAML snippet is a sample of how you would configure your plugin:

```yaml
profiles:
  spinnaker:
    spinnaker:
      extensibility:
        plugins:
          # The plugin id you defined in your build.gradle
          My.Plugin.Id:
            enabled: true
            # It's important to note here that this must be a SemVer compatible
            # string (i.e. do not include a `v` in front of the version string)
            version: "0.1.0"
  gate:
    spinnaker:
      extensibility:
        # This snippet is necessary so that Gate can serve your plugin code to Deck
        deck-proxy:
          enabled: true
          plugins:
            My.Plugin.Id:
              enabled: true
        repositories:
          myPluginRepo:
            enabled: true
            url: https://raw.githubusercontent.com/my-organization/my-plugin-repo/master/repositories.json
```

> After calling `hal deploy apply` or applyingn your `SpinnakerService` manifest

Reviewer 2: s/applyingn/applying

Author: - [x] typo
